Feature: Improve Stream Info Popup

zee-bit opened the pull request:
- [x] Add Stream Members hotkey in README. [merged commit 8ce2edb]
- [x] Add Stream creation date in Stream Info popup [New in Server v4.0, ZFL 30]
- [x] Add Stream email
- [x] bugfix: Fix case when Weekly message count is None
- [x] Add Message retention days [New in server v3.0, ZFL 17]
- [ ] Notification settings [issue #887]

neiljp commented:
@zee-bit We should have included the README in the previous iteration, so it's good to see it here :+1:

For the stream creation date, it's good to see the check vs server feature level here, but note the type of this property, and we'll want to test against a range of values.

Also note that we may want to adjust the model data of the subscriptions to be uniform across different levels (abbreviated to ZFL in places), and then have the UI just use the data that's guaranteed to be present - ie. place the conditional(s) in the model. Ultimately this could be part of a move to store/expose only `StreamData` rather than the underlying data from the server.

zee-bit commented:
@zulipbot add "PR needs review"

neiljp commented on tests/ui_tools/test_popups.py:
```diff
@@ -590,6 +594,7 @@ def test_keypress_navigation(self, mocker, widget_size,
             'subscribers': [],
             'date_created': 1472047124,
             'stream_weekly_traffic': 140,
+            'email_address': 'zt@chat.zulip.org',
```
In practice these email addresses can be very long. A separate test with that kind of example would be useful. For the reason of long emails, I would suggest we consider allowing/forcing the URL onto the next line. For example, try applying your PR to the #announce stream on czo.

zee-bit replied:
I understand your concern. I'll fix this too.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1098,7 +1098,8 @@ def timestamp_to_local_datetime(timestamp: int) -> str:
         stream_email = stream['email_address']
 
-        weekly_msg_count = stream['stream_weekly_traffic']
+        weekly_msg_count = ('N/A' if stream['stream_weekly_traffic'] is None
+                            else str(stream['stream_weekly_traffic']))
```
You're referring to a commit, not a PR, in this commit text.

If `None` is always due to it being just-created, it would be useful to indicate that, and expand N/A to something clearer in any case.

I'd be happy merging this commit sooner, but the tests in the first commit are dependencies of all the others.

zee-bit replied:
Adding this commit in the same PR has led to this. I asked in #zulip-terminal about adding this fix in this PR or another PR, and we decided to go with the former. If you want, I can add this to the PR I mentioned few comments above. This should segregate this fix from the class parameters I declared in the prior commit?

neiljp commented on zulipterminal/model.py:
```diff
@@ -131,6 +131,26 @@ def __init__(self, controller: Any) -> None:
+            if feature_level is None or feature_level < 17:
+                stream['message_retention_days'] = None
+            else:
+                stream['message_retention_days'] = (
+                    message_retention_days_response(
+                        *(self.initial_data['realm_message_retention_days'],
+                          True) if stream['message_retention_days'] is None
+                        else (stream['message_retention_days'], False)
+                    )
+                )
```
This is rather confusing to read. Do we need that nested function? The one-liner if/else is not easy to read.

zee-bit replied:
I'll fix this.
Just to clarify, do you want me to remove the whole function altogether, or just simplify the conditions for calling that function with those parameters?

neiljp reviewed (commented):
@zee-bit This generally looks good :+1: Comments inline.

zee-bit commented on tests/ui_tools/test_popups.py:
```diff
@@ -550,24 +550,62 @@ def test_keypress_navigation(self, mocker, widget_size,
         super_keypress.assert_called_once_with(size, expected_key)
 
 
+@pytest.mark.parametrize('stream, stream_id, server_feature_level', [
```
Yes. I thought it will increase our scope of testing, and potentially detect/eliminate any bug that may get unnoticed, if we tried to localize the parameters inside the class?

zee-bit replied:
I have removed the class parameterization and moved the fixtures to another test function(`test_popup_height`), since the popup height will change depending on the server version running - which corresponds to the numbers of fields supported in that version. :)

zee-bit commented on README.md:
```diff
@@ -196,6 +196,7 @@ source ~/.zshenv
 | Toggle topics in a stream                             | <kbd>t</kbd>                                  |
 | Mute/unmute Streams                                   | <kbd>m</kbd>                                  |
 | Show stream information & modify settings             | <kbd>i</kbd>                                  |
+| Show stream members from stream information box       | <kbd>m</kbd>                                  |
```
So, I'll drop this commit from the PR?

zee-bit commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1079,6 +1081,19 @@ def __init__(self, controller: Any, stream_id: int) -> None:
         self.controller = controller
         stream = controller.model.stream_dict[stream_id]
 
+        def timestamp_to_local_datetime(timestamp: int) -> str:
+            local_timezone = tzlocal.get_localzone()
+            local_time = datetime.fromtimestamp(timestamp, local_timezone)
+
+            formatted_datetime = local_time.strftime("%B %d, %Y")
+            return formatted_datetime
```
Hmm. Seems like the class parametrization was not a very good idea afterall. But, I am not sure how else to add tests to this commit without making the subsequent commits dependent on it.

If you want to centralize this datetime conversion first, I can put it up on a separate PR inside maybe `helper.py` and then refactor this commit to use that function. If you have any other idea in mind, it would help me to know. :)

neiljp replied:
The centralized preferred formatting is merged now as a Model method.

zee-bit commented:
@neiljp PR has been updated.
Using `Overlay` for pyperclip error was looking ugly(and, not in place with our current UI), so I have decided to use the already implemented `NoticeView`, prompting those specific users to install the required utility package for `pyperclip`. Other previous reviews have been addressed as well(except maybe class parameterization, since I wasn't sure if you wanted me to remove those?).

zee-bit commented:
@zulipbot add "PR needs review"
@zulipbot remove "PR needs update"

zee-bit commented:
On second look, I dont think we use Class parameterization anywhere else in our codebase. So, I have removed it, and moved the parameters to another function that tests `stream_info`'s popup height.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1180,6 +1180,13 @@ def __init__(self, controller: Any, stream_id: int) -> None:
 
+        # New in feature level 17, server verison 3.0
+        msg_retention_days = (
+            [("Message retention days", stream["message_retention_days"])]
```
See my other comment regarding extracting another variable.

preetmishra commented on zulipterminal/model.py:
```diff
@@ -156,6 +156,32 @@ def __init__(self, controller: Any) -> None:
+                    stream["message_retention_days"] = message_retention_days_response(
+                        stream["message_retention_days"], False
+                    )
+
```
I think we can simplify the logic and conditionals a bit if we structure this differently.

zee-bit replied:
I have taken the `realm_message_retention_days` field out of the conditional which makes it a bit cleaner, but I don't get how it can be simplified further. It already looks pretty straightforward?

Could you please explain what structure you had in mind?

preetmishra replied:
Yeah, that makes it better. Additionally, you could do something like:
```python3
                    stream["message_retention_days"] = message_retention_days_response(
                        realm_message_retention_days, stream["message_retention_days"] is None
                    )
```
That would simplify the `else` block.

zee-bit replied:
I believe you are talking about the first else block?
The conditional within that have `stream['message_retention_days']` and `realm_message_retention_days` as args to the function call. I don't think they will have the same value at any point of time?

zee-bit replied:
@neiljp All of the above makes sense. Just had one small doubt; for server versions < 17 do we need to include this field in the `Subscription` data?

zee-bit replied:
If we include this field with a value of `None`, as before, then it'll get mixed with the special `None` value (denoting that it inherits the organization-level setting) since we are not over-writing this field.

zee-bit replied:
Yes, this is resolved with the `cached_retention_text` and the idea to fallback to the realm value on older server versions.

preetmishra reviewed (commented):
@zee-bit Thanks for working on this! This seems to work well locally. :+1:

The changes look good overall. Though, since the PR is a bit dated, it would benefit from the skills that you've acquired over the past weeks. :) I have left comments in-line.

zee-bit commented on tests/ui_tools/test_popups.py:
```diff
@@ -755,7 +755,7 @@ def test_keypress_stream_members(self, mocker, key, widget_size):
                     "rendered_description": "hey",
                     "subscribers": [],
                     "date_created": None,
-                    "stream_weekly_traffic": 123,
```
As per the commit description, if a stream is newly created (within a week), the `stream_weekly_traffic` would be `None` as per the API docs [here](https://zulip.com/api/get-subscriptions#response).

I was not sure if I should add new tests with this field as `None` or just change the pre-existing tests (to prevent adding unnecessary tests and keep the commit smaller). Let me know if adding new tests is better. :)

The latter did not break the tests because the tests check for popup height, which remains the same even if the field is `None` cause we're displaying an appropriate message in that case.

zee-bit replied:
I have added a new fixture with `stream_weekly_traffic` as `None`, without modifying the existing ones.

zee-bit commented:
@preetmishra PR updated. :)

zee-bit commented on zulipterminal/core.py:
```diff
@@ -338,6 +339,29 @@ def stream_muting_confirmation_popup(self, button: Any) -> None:
+        try:
+            pyperclip.copy(stream_email)
+            clipboard_text = pyperclip.paste()
+            assert clipboard_text == stream_email
```
It's difficult to test this since we have an external module involved which makes use of system packages (`xclip`/`xsel` on Linux, and others on different platforms). We can't know the platform on which our tests are running, and if those packages are installed in the system, nor if the text was successfully copied to clipboard.

We can test for the exception case using `side_effects` though, but I don't think that'd be any useful?

preetmishra replied:
I agree with Neil. That is what I was talking about in the first statement.

Re testing, we don't have to test the external module. We could just test whether we're calling the external module correctly.

neiljp reviewed (commented):
@zee-bit This looks useful and functionally appears to work well :+1: (the copy seems to work fine for me!)

I think if we can simplify the tests then this will be a lot easier to work with - and the tests you add ea rly on block the later commits, which I think we discussed previously. So, once they are clean a lot should be mergeable, pending my other feedback.

zee-bit commented on zulipterminal/core.py:
```diff
@@ -436,6 +437,31 @@ def stream_muting_confirmation_popup(
+                    "\nYou can fix this by installing any ONE of the "
+                    "copy/paste mechanisms using:\n\n",
+                ),
+                ("msg_code", "sudo apt-get install xclip [Recommended]\n"),
+                ("msg_code", "sudo apt-get install xsel"),
```
The first 2 (xclip, xsel) are system packages whereas the other 2 (gtk, PyQt4) are python modules, which still needs the system packages to run so I'm not sure why they are listed in the website. The copying also seems to work fine without these python modules installed.

preetmishra commented:
@zee-bit Thanks for the update on the feedback. This looks pretty good already! I have left replies to the respective in-line comments. :+1:

Let's work on simplifying the tests as Neil mentioned before moving ahead with the next review iteration.

zee-bit commented:
@neiljp @preetmishra I have updated this PR as per both of your reviews.
Re the `stream_data_to_vary`, I have defined and used the `general_stream` fixture from conftest, of which I'm not sure if it is a bad decision. Let me know if I should instead define a stream data dict in the main test and then build up from fixtures.

neiljp reviewed (commented):
@zee-bit This addresses a lot of my structural points, probably most left are the details :+1:

The use of the fixture looks good to me, certainly for now. We may want to follow-up with private variations and similar later, but we can build on this.

zee-bit commented on zulipterminal/core.py:
```diff
@@ -441,6 +442,29 @@ def stream_muting_confirmation_popup(
+            if clipboard_text == text:
+                self.report_success(f"{text_category} copied successfully!")
+            else:
+                self.report_error(f"Could not copy {text_category}. Please try again.")
```
I was not sure what to mention as a possible *try*/*hint*, since failing to copy would mean it's an internal `pyperclip` issue. Re-thinking now, I believe we could mention something along the lines of: "Could not copy Test. Error in the dependency library" or just remove the second sentence altogether?

neiljp replied:
Possibly worth rewording and avoiding the 'try again' sentence - after all, we *have* copied, we can just warn (strongly) that a pasted value may be incorrect, or something like that? That's more transparent to the user - conceptually they may have some special clipboard system which could upset this, for example? Then they'd know it was copied but a verification didn't work.

preetmishra reviewed (commented):
@zee-bit Thanks for the updates! :+1: This continues to work well. I've left some in-line comments.

zee-bit commented on zulipterminal/core.py:
```diff
@@ -441,6 +442,31 @@ def stream_muting_confirmation_popup(
+                self.report_success(f"{text_category} copied successfully")
+            else:
+                self.report_warning(
+                    f"{text_category} copied, but the clipboard text does not match"
```
This might happen if there is an internal issue with the `pyperclip` module. Neil also explained other possible trigger conditions in [this](https://github.com/zulip/zulip-terminal/pull/880#discussion_r682727216) comment.

zee-bit commented on zulipterminal/core.py:
```diff
@@ -441,6 +442,31 @@ def stream_muting_confirmation_popup(
+        try:
+            pyperclip.copy(text)
+            clipboard_text = pyperclip.paste()
+            if clipboard_text == text:
```
This was done to address your previous review [here](https://github.com/zulip/zulip-terminal/pull/880#discussion_r679679485), where we decided to not use `assert` statements but simple conditional to check the correctness of the copied text.

preetmishra replied:
Ah, I understand now. I was under the impression that the module would copy correctly or not throw an exception.

neiljp commented on tests/conftest.py:
```diff
@@ -192,45 +192,52 @@ def logged_on_user():
-    "email_address": "general@example.comm",
-    "subscribers": [1001, 11, 12],
-}
+@pytest.fixture
```
Some appear to have just shifted to the next commit; let's address these in a different PR?

neiljp commented:
@zee-bit For the record, I just merged the first 7 commits as 0e3d91e6bf77986c8fccb7d520412cfd0de777c2 and those prior :tada:

This should allow focus on the last commit by itself :+1:

neiljp commented on tests/model/test_model.py:
```diff
@@ -220,6 +220,70 @@ def test_register_initial_desired_events(self, mocker, initial_data):
+
+        model.normalize_and_cache_message_retention_text()
+
+        for key in to_vary_in_stream_dict:
+            assert model.cached_retention_text[key] == expect_msg_retention_text[key]
```
Minor: "key" is very broad; these are stream ids, right?

neiljp commented on tests/model/test_model.py:
```diff
@@ -220,6 +220,70 @@ def test_register_initial_desired_events(self, mocker, initial_data):
+            "expect_msg_retention_text",
+        ],
+        [
+            case(
+                {1: {}},
+                None,
+                10,
+                {1: "Indefinite [Organization default]"},
+                id="ZFL=None_no_stream_retention_realm_retention=None",
```
Ah, this is the `None` case that I queried elsewhere, where the organization value slips through to the stream level.

Was `None` a realm value, though? I assume it followed the original rule, ie. -1 for indefinite and +n for days, so I'm not sure about None?

The #api-design topic doesn't seem to have clarified this exactly, so perhaps Mateusz could answer in a follow-up or you could refer him here?

zee-bit replied:
Right, this is why we were handling the `days == None` case earlier.
To clarify, `None` was a realm value earlier (between server v2-3, as per my testing on the dev server) and it had the same meaning as the current `-1` for indefinite which was changed by @mateuszmandera's [commit](https://github.com/zulip/zulip/commit/7a03e2a7fe259a3410cadbb1a9a4955aa1fae3c7). So to summarize, I believe this was how the server's representation changed over time:
- Server version 2-3: realm value could have `None` for indefinite and +n for days
- Server version 3.0: added stream message retention days, with `None` for org default, `-1` for indefinite and +n for days
- Server version 3+: realm's `None` value was changed to `-1` by Mateusz's above commit.

(I hope Mateusz should get notification from the above mention, still I'll refer him here from the stream to confirm my hypothesis)

mateuszmandera replied:
Ah, that was an interesting dive through old commits :grinning: This overview sounds correct to me based on that. To be a bit more precise, `message_retention_days` was a thing for streams before 3.0 but only in the sense of the backend logic - but it indeed wasn't exposed via the API in any way (the value had to be set manually via `manage.py shell` or db query). I presume that `zulip-terminal` only cares about the API history here though, where indeed this is only a thing since 3.0 - https://github.com/zulip/zulip/commit/c488a35f10ac7b1e5e20fc46302c85ce2364e825

neiljp reviewed (commented):
@zee-bit This last commit is mainly minor points since we've discussed a lot, but the early Zulip version behavior still seems unclear. Other than the existing commit being quite large, having the text extraction in a separate commit will help with focusing anyone else on how we're handling it, without having display code in the way.

zee-bit commented on zulipterminal/model.py:
```diff
@@ -192,6 +194,37 @@ def __init__(self, controller: Any) -> None:
         self._start_presence_updates()
 
+    def message_retention_days_response(self, days: int, org_default: bool) -> str:
+        suffix = " [Organization default]" if org_default else ""
+        return ("Indefinite" if (days == -1 or days is None) else str(days)) + suffix
```
I think this should be already answered by the below comment of mine and Mateusz?

neiljp commented:
@zee-bit Thanks for getting this final commit work done :+1: This now passes CI after the pytz fixup. I'm going to merge this now :tada: (we can follow-up with other features in separate PRs - could you file issues for what we discussed?)

@mateuszmandera Thanks for confirmation of the findings - do we want to improve the API docs for this?
